iwyu_test_util.py

```diff
@@ -419,6 +436,112 @@ def _GetLaunchArguments(cc_file):
   return shlex.split(args)
 
 
+def _CheckForCompileTimeMacro(macro):
+    """This assumes that the local clang++ and the include-what-you-use
+    under test are using the same libclang, and therefore the same default libraries.
+
+    Look for the specified macro in the clang output.
+    """
+    txt = '#include <cstddef>'
+
+    result = subprocess.run(['clang', '-dM', '-E', '-x', 'c++', '-'], capture_output=True,
+                            input=txt, text=True)
+
+    return " "+macro+" " in str(result.stdout)
+
+
+def _CheckExtraCommandLine(args):
+  """Check if particular options are on the command line"""
+  env_iwyu_extra_args = os.getenv('IWYU_EXTRA_ARGS')
+  iwyu_extra_args = shlex.split(env_iwyu_extra_args)
+  standardised_args = ' '.join(iwyu_extra_args)
+  return args in standardised_args
+
+
+def _EvalCondition(cond):
+  """Evaluate the input condition.
+
+  All conditions comprise a label and a single argument. The supported
+  tests are:
+
+  IwyuHasTargetSupportFor arg - whether include-what-you-use supports
+                                the architecture 'arg' as a target.
+
+  TestingCXXSTL (libc++|libstdcpp|canonical) - whether the current
+                                test is compiling against a particular
+                                C++ standard library.
+  """
+  # Some helper functions to handle conditions
+  def isTestingLibCxx():
+    return _CheckExtraCommandLine('-stdlib=libc++') or _CheckForCompileTimeMacro('_LIBCPP_VERSION')
+  def isTestingLibstdcpp():
+    return _CheckExtraCommandLine('-stdlib=libstdc++') or _CheckForCompileTimeMacro('__GLIBCXX__')
+  def isTestingCanonicalClangMappings():
+    return _CheckExtraCommandLine('-Xiwyu --experimental=clang_mappings')
+  def iwyuHasTargetSupportFor(tgt):
+    iwyu = _GetIwyuPath()
+    result = subprocess.run([iwyu, '--print-targets'], capture_output=True)
+    return tgt in str(result.stdout)
+
+  m = _IWYU_COND.match(cond)
+  if m:
+    label = m.group(1)
+    arg = m.group(2)
+
+    if label == 'IwyuHasTargetSupportFor':
+      return iwyuHasTargetSupportFor(arg)
+    elif label == 'TestingCXXSTL':
```

Comment:
I don't really understand what this label is saying, particularly not with the `canonical` mapping mixed in.

Could you maybe split these features up into two:

- `stdlib {libcxx|libstdcxx}`
- `builtin-mappings {canonical|default?|...?}`

It's not 100% clear to me what the mappings tag would indicate, so maybe let's cut it for now and use the target feature as a motivating example?